route-rs-afpacket/tests/integration_test.rs

```diff
@@ -0,0 +1,174 @@
+        assert!(sock.is_ok());
+        sock.unwrap()
+    };
+
+    sock.split()
+}
+
+fn make_packet(rng: &mut impl Rng) -> Vec<u8> {
+    let body = {
+        let mut body = vec![0; 64];
+        rng.fill(&mut body[..]);
+        body
+    };
+    let mut udp_pkt = packets::UdpSegment::empty();
+    udp_pkt.set_src_port(3001);
+    udp_pkt.set_dest_port(3002);
+    udp_pkt.set_payload(&body);
+    let mut ipv4_pkt = packets::Ipv4Packet::encap_udp(udp_pkt);
+    ipv4_pkt.set_src_addr(net::Ipv4Addr::new(10, 0, 0, 1));
+    ipv4_pkt.set_dest_addr(net::Ipv4Addr::new(10, 0, 0, 2));
+    ipv4_pkt.set_ttl(2);
+    let mut eth_pkt = packets::EthernetFrame::encap_ipv4(ipv4_pkt);
+    eth_pkt.set_src_mac(packets::MacAddr::new([0xac, 0x67, 0x3d, 0xfa, 0xbc, 0xdd]));
+    eth_pkt.set_dest_mac(packets::MacAddr::new([0xac, 0x67, 0x3d, 0xfa, 0xbc, 0xdf]));
+
+    eth_pkt.data
+}
+
+#[test]
+#[should_panic]
+#[ignore]
+fn output_panics_when_built_without_ingressor() {
+    let (s, _) = create_socket();
+
+    AfPacketOutputLink::new().channel(s).build_link();
+}
+
+#[test]
+#[should_panic]
+fn output_panics_when_built_without_channel() {
+    let packet_generator = immediate_stream(vec![]);
+
+    AfPacketOutputLink::new()
+        .ingressor(packet_generator)
+        .build_link();
+}
+
+#[test]
+#[should_panic]
+#[ignore]
+fn output_panics_when_built_with_multiple_ingressors() {
+    let (s, _) = create_socket();
+    let pkt_gen_1 = immediate_stream(vec![]);
+    let pkt_gen_2 = immediate_stream(vec![]);
+
+    AfPacketOutputLink::new()
+        .ingressors(vec![pkt_gen_1, pkt_gen_2])
+        .channel(s)
+        .build_link();
+}
+
+#[test]
+#[should_panic]
+fn input_panics_when_built_with_ingressors() {
+    AfPacketInputLink::new()
+        .ingressors(vec![immediate_stream(vec![])])
+        .build_link();
+}
+
+#[test]
+#[should_panic]
+fn input_panics_when_built_without_channel() {
+    AfPacketInputLink::new().build_link();
+}
+
+#[test]
+#[ignore]
+fn input_identity() {
+    let _timeout = Duration::from_secs(1);
+    let mut rt = initialize_runtime();
+
+    // make packet
+    let mut rng = rand::thread_rng();
+    let pkt_data = make_packet(&mut rng);
+    let ext_pkt_data = pkt_data.clone();
+
+    let results = rt.block_on(async {
+        let iface_name = CString::new("lo").unwrap();
+
+        // set up router
+        let mut router_side = afpacket::AsyncBoundSocket::from_interface(&iface_name).unwrap();
+        router_side.set_promiscuous(true).unwrap();
+        let (_, rx_rtr) = router_side.split();
+        let mut external = afpacket::AsyncBoundSocket::from_interface(&iface_name).unwrap();
+
+        // set up link
+        let link = AfPacketInputLink::new().channel(rx_rtr).build_link();
+        let _external_task = tokio::spawn(async move {
+            // send packet from afpacket socket -> route-rs-afpacket
+            external.send(&ext_pkt_data).await.unwrap();
+        });
+
+        run_link(link).await
```

Comment:
change to `test_link(link, timeout).await` I'll merge the new test_link() for you by EoW so you can merge this.